crates/typst/src/eval/version.rs

```diff
@@ -0,0 +1,223 @@
+use crate::diag::{bail, StrResult};
+use crate::eval::Value;
+use ecow::EcoVec;
+use serde::{Serialize, Serializer};
+use std::cmp::Ordering;
+use std::fmt::{Debug, Display, Formatter};
+use std::hash::{Hash, Hasher};
+use std::iter::repeat;
+use std::str::FromStr;
+use typst::eval::Array;
+
+/// A version, with any number of components.
+///
+/// The list of components is semantically extended by an infinite list of zeros.
+/// This means that, for example, `0.8` is the same as `0.8.0`.
+/// As a special case, the empty version (that has no components at all) is the same as `0`, `0.0`, `0.0.0`, and so on.
+#[derive(Clone)]
+pub struct Version(EcoVec<VersionComponent>);
+
+// helper macro to make sure COMPONENT_NAMES and resolve_component_name never go out of sync
+macro_rules! component_names {
```

Comment:
I try hard to avoid macros unless absolutely necessary. Is `named_components` even used?

Reply:
Both of the items that this macro generates are used and it would be really bad if they ever went out of sync.
That's why I chose to make this a macro, which makes that issue impossible.

Reply:
Couldn't you remove `resolve_component_name` in favor of a simple `Self::COMPONENT_NAMES.iter().position(v| v == name)`?

Reply:
Hm, yeah I guess that's good enough. You caught me red-handed with premature optimization...

Reply:
It's really not on a hot path, so it's fine.